bpf: support ipv6 egressgateway policies

rgo3 opened the pull request:
This PR enables ipv6 egressgateway policies at the datapath level, ensuring that IPv6 traffic can be matched when respective policies are in place. It should be noted that intra cluster traffic that needs to be redirected to the correct gateway node, is still going through an IPv4 tunnel.

Apart from that, all the newly IPv6 related changes in this commit should be akin to the already existing code paths that provide the egressgateway functionality for IPv4 traffic.

It also adds the same tests as we currently have for ipv4 policies.

rgo3 committed:
bpf: support ipv6 egressgateway policies

This commit enables ipv6 egressgateway policies at the datapath level,
ensuring that IPv6 traffic can be matched when respective policies are
in place. It should be noted that intra cluster traffic that needs to be
redirected to the correct gateway node, is still going through an IPv4
tunnel.
Apart from that, all the newly IPv6 related changes in this commit
should be akin to the already existing code paths that provide the
egressgateway functionality for IPv4 traffic.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

rgo3 commented on bpf/lib/egress_gateway.h:
```diff
@@ -260,4 +443,42 @@ int egress_gw_handle_packet(struct __ctx_buff *ctx,
+        return CTX_ACT_OK;
+
+    /* Send the packet to egress gateway node through a tunnel. */
+    return __encap_and_redirect_with_nodeid(ctx, &gateway_ip,
```
@julianwiedmann This currently complains about the `gateway_ip` variable being of type `union v6addr` because `__encap_and_redirect_with_nodeid` expects a v4 address here. Do I need an additional call to `lookup_ip6_remote_endpoint` to retrieve the tunnel endpoint IP for the ipv6 gateway ip address?

rgo3 replied:
Or does the gateway_ip continue to be an ipv4 address even with ipv6 policies?

julianwiedmann replied:
> Or does the gateway_ip continue to be an ipv4 address even with ipv6 policies?

yes! Supporting IPv6 node addresses would be a separate effort, right now we can focus on a dual-stack environment with IPv4 node-to-node connectivity. So for now all the gateway IPs would be IPv4.

rgo3 replied:
Thanks, I'll go fix that.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1559,7 +1559,9 @@ int cil_to_netdev(struct __ctx_buff *ctx)
 	{
 		void *data, *data_end;
 		struct iphdr *ip4;
-		struct ipv4_ct_tuple tuple = {};
```
This code path handles the pod-to-world direction, on the worker node. Matching against EGW policy, and potentially redirecting (via VXLAN) to the gateway node.

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -149,6 +149,36 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
+		__u32 egress_ifindex = 0;
+		union v6addr snat_addr, daddr;
+
+		daddr = ip6->daddr;
```
This grabs the pod-to-world traffic as it exits the overlay network on the gateway node. Does *another* EGW policy lookup to obtain the desired `egressIP`, does a FIB lookup (to get the corresponding network interface), and redirects straight to that interface.

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -476,6 +476,17 @@ struct egress_gw_policy_entry {
+struct egress_gw_policy_entry6 {
+    union v6addr gateway_ip;
+    union v6addr egress_ip;
+};
```
Can you please block off an additional `__u32 reserved` here? I have future plans :slightly_smiling_face:

rgo3 replied:
Was going to make a comment for this either way since I am turning the gateway_ip var into a __u32. Do we then want event more reserved bytes?

julianwiedmann replied:
Blocking off some space to store an IPv6 gateway address can't hurt I suppose :)

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1514,6 +1521,27 @@ snat_v6_needs_masquerade(struct __ctx_buff *ctx __maybe_unused,
+
+		return NAT_NEEDED;
+	}
+#endif
```
This is in the egress path on the gateway node, as the packet (local or redirected in `from-overlay`) ends up in to-netdev on the egress interface.. We apply SNAT with the desired `egressIP`.

julianwiedmann commented on bpf/lib/nodeport_egress.h:
```diff
@@ -84,6 +84,24 @@ static __always_inline int nodeport_snat_fwd_ipv6(struct __ctx_buff *ctx,
+		ret = egress_gw_fib_lookup_and_redirect_v6(ctx, target.addr,
+							tuple.daddr, target.ifindex,
+							ext_err);
+		if (ret != CTX_ACT_OK)
```
This is used during masquerading, for traffic by a client that's local on the gateway node. When it hits `to-netdev` (based on the default route) and the masquerading code sees that the desired `egressIP` lives on a **different** network interface, we bounce the packet over to the correct network interface (and then do the masquerading there).

julianwiedmann reviewed (commented):
Thank you! I just took a first pass to sketch out the packet flow, so we're aligned on what those parts do. It *looks* complete :slightly_smiling_face:.

One subtle thing to check is whether when masquerading (on the gateway node), we already pass back the correct CTX_ACT_REDIRECT (when bouncing the packet to a different egress interface) and `orig_addr` for `send_trace_notify6()`. I remember taking some shortcuts back then, because we only needed it for IPv4 ...

rgo3 committed:
bpf/tests: add egress gw tests for ipv6

This commit provides a translation of the existing IPv4 BPF integration
test suites for egressgateway policies into IPv6 equivalent tests.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

julianwiedmann commented on bpf/tests/lib/egressgw.h:
```diff
@@ -271,3 +278,221 @@ static __always_inline int create_ct_entry(struct __ctx_buff *ctx, __be16 client
+{
+	union v6addr ext_svc_ip = EXTERNAL_SVC_IP_V6;
+	union v6addr client_ip = CLIENT_IP_V6;
+	struct ipv6_ct_tuple tuple = {};
```
```suggestion
	struct ipv6_ct_tuple tuple __align_stack_8 = {};
```

rgo3 commented on bpf/tests/lib/egressgw_policy.h:
```diff
@@ -22,11 +22,46 @@ static __always_inline void add_egressgw_policy_entry(__be32 saddr, __be32 daddr
+#ifdef ENABLE_IPV6
+static __always_inline void add_egressgw_policy_entry_v6(const union v6addr *saddr,
+							 const union v6addr *daddr,
+							 __u8 cidr,
```
Should this be called `prefix` instead of `cidr` for ipv6 addresses?

rgo3 replied:
I'll leave it as is.

rgo3 commented on bpf/tests/tc_egressgw_redirect_from_host.c:
```diff
@@ -213,6 +216,214 @@ int egressgw_drop_no_egress_ip_check(const struct __ctx_buff *ctx)
+
+	ipcache_v6_add_entry(&ext_svc_ip, 0, WORLD_IPV6_ID, 0, 0);
+	create_ct_entry_v6(ctx, client_port(TEST_REDIRECT));
+	add_egressgw_policy_entry_v6(&client_ip, &ext_svc_ip, 48, GATEWAY_NODE_IP, &egress_ip);
```
Besides me being inconsistent using /48 and /64 prefixes in the different egw BPF testsuites :grimacing:, I think I should change all of them to /64. IPv4 testsuites mostly use /24, not sure what the best equivalent is.

julianwiedmann replied:
I wouldn't sweat it. If you add a macro & use it in all the entries, it's easy enough to change later if needed.

rgo3 commented on bpf/lib/nodeport_egress.h:
```diff
@@ -701,10 +717,12 @@ handle_nat_fwd(struct __ctx_buff *ctx, __u32 cluster_id, __u32 src_id,
+							    is_defined(ENABLE_IPV6)),
+						      __and(is_defined(ENABLE_HOST_FIREWALL),
+							    is_defined(IS_BPF_HOST)),
+						      __and(is_defined(ENABLE_EGRESS_GATEWAY_COMMON),
```
This currently throws a checkpatch error, however, the ipv4 code looks exactly the same so I thought we might be able to ignore that given that checkpatch is not required to pass.

julianwiedmann replied:
sgtm

rgo3 replied:
@ysksuzuki discussion of why we accepted the checkpatch failure.

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -150,6 +150,36 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
+			if (ipv6_addr_equals(&snat_addr, &EGRESS_GATEWAY_NO_EGRESS_IP_V6))
+				return DROP_NO_EGRESS_IP;
+
+			ret = ipv6_l3(ctx, ETH_HLEN, NULL, NULL, 0);
```
```suggestion
			ret = ipv6_l3(ctx, ETH_HLEN, NULL, NULL, METRIC_EGRESS);
```

rgo3 replied:
Could you elaborate why this is necessary? I can't find (or am too blind to see) the ipv4 equivalent.

julianwiedmann replied:
Under the hood, we're decrementing the packet's TTL - and drop it if the resulting TTL is 0.

For IPv4 the calling code then throws a drop notification. But for IPv6 we branch off into a tailcall that attempts to return a `TIME_EXCEEDED` ICMP error message. If that fails, it raises a drop notification from inside the tailcall, and needs to know which direction it should report. So for IPv6, `ipv6_l3()` takes the `direction` parameter.

But to correct myself - from the surrounding code and overall path (the dropped packet was in the ingressing path), `METRIC_INGRESS` is probably a better fit here.

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -150,6 +150,36 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
+
+		if (egress_gw_snat_needed_hook_v6((union v6addr *)&ip6->saddr,
+						  (union v6addr *)&ip6->daddr, &snat_addr,
+						  &egress_ifindex)) {
```
```suggestion
                ipv6_addr_copy(&daddr, ip6->daddr);
		if (egress_gw_snat_needed_hook_v6((union v6addr *)&ip6->saddr,
						  &daddr, &snat_addr,
						  &egress_ifindex)) {
```

This is probably also needed to make the FIB lookup work as expected (where we pass `daddr`).

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -486,6 +486,18 @@ struct egress_gw_policy_entry {
+struct egress_gw_policy_entry6 {
+	__u32 gateway_ip;
+	union v6addr egress_ip;
+	__u32 reserved[3]; /* reserved for future extension, e.g. v6 gateway_ip */
```
```suggestion
	union v6addr egress_ip;
	__u32 gateway_ip;
	__u32 reserved[3]; /* reserved for future extension, e.g. v6 gateway_ip */
```

This should help with struct-internal alignment.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1533,6 +1540,27 @@ snat_v6_needs_masquerade(struct __ctx_buff *ctx __maybe_unused,
 
+/* Check if the packet matches an egress NAT policy and so needs to be SNAT'ed.
+ *
+ * This check must happen before the IPV4_SNAT_EXCLUSION_DST_CIDR check below as
```
```suggestion
 * This check must happen before the IPV6_SNAT_EXCLUSION_DST_CIDR check below as
```

Alternatively just drop the comment, when it's otherwise identical to the IPv4 path?

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -921,14 +922,27 @@ nodeport_rev_dnat_ingress_ipv6(struct __ctx_buff *ctx, struct trace_ctx *trace,
+	 */
+	if (egress_gw_reply_needs_redirect_hook_v6(ip6, &tunnel_endpoint, &dst_sec_identity)) {
+		trace->reason = TRACE_REASON_CT_REPLY;
+		src_sec_identity = WORLD_ID;
+		goto encap_redirect;
+	}
+#endif /* ENABLE_EGRESS_GATEWAY_COMMON */
+
 	return CTX_ACT_OK;
 
-#ifdef TUNNEL_MODE
+#if (defined(ENABLE_EGRESS_GATEWAY_COMMON) && (defined(IS_BPF_XDP) || defined(IS_BPF_HOST))) ||	\
+    defined(TUNNEL_MODE)
 encap_redirect:
 	src_port = tunnel_gen_src_port_v6(&tuple);
 
 	ret = nodeport_add_tunnel_encap(ctx, IPV4_DIRECT_ROUTING, src_port,
-					tunnel_endpoint, REMOTE_NODE_ID, dst_sec_identity,
+					tunnel_endpoint, src_sec_identity, dst_sec_identity,
```
Can we preserve the `REMOTE_NODE_ID` for LB reply traffic, same as in the IPv4 path?

julianwiedmann commented on bpf/tests/lib/egressgw.h:
```diff
@@ -271,3 +279,224 @@ static __always_inline int create_ct_entry(struct __ctx_buff *ctx, __be16 client
+	l2 = pktgen__push_ethhdr(&builder);
+	if (!l2)
+		return TEST_ERROR;
+	l2->h_proto = bpf_htons(ETH_P_IPV6);
```
oh, did you need that? It should be handled by the pkt builder, when it iterates over the layers during `pktgen__finish()`.

rgo3 replied:
It is handled by the pkt builder. I had some problems when initially developing the test and added these statements while something else was my culprit. Consider this removed.

julianwiedmann commented on bpf/tests/lib/egressgw.h:
```diff
@@ -271,3 +279,224 @@ static __always_inline int create_ct_entry(struct __ctx_buff *ctx, __be16 client
+		ipv6_addr_copy((union v6addr *)&l3->saddr, &client_ip);
+		ipv6_addr_copy((union v6addr *)&l3->daddr, &ext_svc_ip);
+	}
+	l3->nexthdr = IPPROTO_TCP;
```
same here, isn't the pkt builder taking care of this?

julianwiedmann reviewed (commented):
Thank you, looks close to ship already!

Took a first swing over the production code, will come back for the tests later :)

julianwiedmann reviewed (approved):
Tests lgtm as well, thank you!
